File: web/src/lib/crypto/keys.ts
import nacl from 'tweetnacl';
import { encodeBase64, decodeBase64 } from 'tweetnacl-util';
import type { KeyPair } from '@/types/chat';

/**
 * ECDH 키쌍 생성 (Curve25519)
 */
export function generateKeyPair(): KeyPair {
  const keyPair = nacl.box.keyPair();
  return {
    publicKey: keyPair.publicKey,
    secretKey: keyPair.secretKey,
  };
}

/**
 * ECDH 공유 비밀 계산
 * 상대방의 공개키와 내 비밀키로 공유 비밀 생성
 */
export function computeSharedSecret(
  theirPublicKey: Uint8Array,
  mySecretKey: Uint8Array
): Uint8Array {
  return nacl.box.before(theirPublicKey, mySecretKey);
}

/**
 * 비밀번호에서 이중 키 유도 (PBKDF2)
 * - 앞 32바이트: authKey (서버 검증용)
 * - 뒤 32바이트: encryptionSeed (암호화 키 시드, 서버 모름)
 */
export async function deriveKeysFromPassword(
  password: string,
  roomId: string
): Promise<{ authKey: Uint8Array; encryptionSeed: Uint8Array }> {
  const encoder = new TextEncoder();
  const keyMaterial = await crypto.subtle.importKey(
    'raw',
    encoder.encode(password),
    'PBKDF2',
    false,
    ['deriveBits']
  );

  const derivedBits = await crypto.subtle.deriveBits(
    {
      name: 'PBKDF2',
      salt: encoder.encode(`blip-room-${roomId}`),
      iterations: 100000,
      hash: 'SHA-256',
    },
    keyMaterial,
    512 // 64바이트 = 512비트
  );

  const derived = new Uint8Array(derivedBits);
  return {
    authKey: derived.slice(0, 32),
    encryptionSeed: derived.slice(32, 64),
  };
}

/**
 * authKey를 SHA-256 해시 (서버 저장용)
 */
export async function hashAuthKey(authKey: Uint8Array): Promise<string> {
  const hashBuffer = await crypto.subtle.digest('SHA-256', authKey.buffer as ArrayBuffer);
  return encodeBase64(new Uint8Array(hashBuffer));
}

/**
 * 공개키를 base64 문자열로 변환
 */
export function publicKeyToString(publicKey: Uint8Array): string {
  return encodeBase64(publicKey);
}

/**
 * base64 문자열을 공개키로 변환
 */
export function stringToPublicKey(str: string): Uint8Array {
  return decodeBase64(str);
}

File: mobile/lib/core/crypto/keys.dart
import 'dart:convert';

import 'package:crypto/crypto.dart' as hash;
import 'package:pinenacl/x25519.dart';
import 'package:pointycastle/export.dart' hide PrivateKey, PublicKey;

import '../constants/app_constants.dart';

/// ECDH 키쌍 (Curve25519)
class BlipKeyPair {
  final Uint8List publicKey;
  final Uint8List secretKey;

  const BlipKeyPair({required this.publicKey, required this.secretKey});
}

/// PBKDF2 유도 결과
class DerivedKeys {
  final Uint8List authKey;
  final Uint8List encryptionSeed;

  const DerivedKeys({required this.authKey, required this.encryptionSeed});

  /// authKey/encryptionSeed 메모리 제로화
  void dispose() {
    authKey.fillRange(0, authKey.length, 0);
    encryptionSeed.fillRange(0, encryptionSeed.length, 0);
  }
}

/// ECDH 키쌍 생성 (web: nacl.box.keyPair)
BlipKeyPair generateKeyPair() {
  final pk = PrivateKey.generate();
  return BlipKeyPair(
    publicKey: Uint8List.fromList(pk.publicKey),
    secretKey: Uint8List.fromList(pk),
  );
}

/// ECDH 공유 비밀 계산 (web: nacl.box.before)
/// pinenacl Box 생성 시 내부적으로 crypto_box_beforenm 호출
Uint8List computeSharedSecret(
    Uint8List theirPublicKey, Uint8List mySecretKey) {
  final box = Box(
    myPrivateKey: PrivateKey(mySecretKey),
    theirPublicKey: PublicKey(theirPublicKey),
  );
  return Uint8List.fromList(box.sharedKey);
}

/// 비밀번호에서 이중 키 유도 (PBKDF2)
/// web: crypto.subtle.deriveBits(PBKDF2, SHA-256, 100000, 512)
/// salt: 'blip-room-{roomId}' (UTF-8)
DerivedKeys deriveKeysFromPassword(String password, String roomId) {
  final salt = utf8.encode('${AppConstants.pbkdf2SaltPrefix}$roomId');
  final passwordBytes = utf8.encode(password);

  final derivator = KeyDerivator('SHA-256/HMAC/PBKDF2')
    ..init(Pbkdf2Parameters(
      Uint8List.fromList(salt),
      AppConstants.pbkdf2Iterations,
      AppConstants.pbkdf2OutputBytes,
    ));

  final derived = derivator.process(Uint8List.fromList(passwordBytes));

  return DerivedKeys(
    authKey: Uint8List.fromList(derived.sublist(0, 32)),
    encryptionSeed: Uint8List.fromList(derived.sublist(32, 64)),
  );
}

/// authKey → SHA-256 → Base64 (서버 검증용)
/// web: crypto.subtle.digest('SHA-256', authKey) → encodeBase64
String hashAuthKey(Uint8List authKey) {
  final digest = hash.sha256.convert(authKey);
  return base64Encode(Uint8List.fromList(digest.bytes));
}

/// 공개키 → Base64 문자열
String publicKeyToString(Uint8List publicKey) {
  return base64Encode(publicKey);
}

/// Base64 문자열 → 공개키
Uint8List stringToPublicKey(String str) {
  return base64Decode(str);
}
